nasl/nasl_frame_forgery.c

```diff
@@ -173,3 +188,225 @@ nasl_send_arp_request (lex_ctxt *lexic)
+  int frame_sz = get_var_size_by_name (lexic, "frame");
+  int use_pcap = get_int_var_by_name (lexic, "pcap_active", 1);
+  int to = get_int_var_by_name (lexic, "pcap_timeout", 5);
+  char *filter = get_str_var_by_name (lexic, "pcap_filter");
+  int ifindex;
+  u_char *answer = NULL;
+  int answer_sz = 0;
+  int bpf = -1;
+  int frame_and_payload = 0;
+
+  // Create the raw socket
+  soc = socket (AF_PACKET, SOCK_RAW, htons (ETH_P_ALL));
+  if (soc == -1)
+    {
+      nasl_perror (lexic, "%s: %s", __func__, strerror (errno));
+      return NULL;
+    }
+
+  // We will need the eth index. We get it depending on the target's IP..
+  if (get_iface_index (ipaddr, &ifindex) < 0)
+    {
+      nasl_perror (lexic, "%s: Missing interface index\n", __func__);
+      return NULL;
+    }
+
+  // Preapre sockaddr_ll. This is necessary for further captures
+  unsigned char dst_haddr[ETHER_ADDR_LEN];
+  memcpy (&dst_haddr, (struct pseudo_frame *) frame, ETHER_ADDR_LEN);
+
+  struct sockaddr_ll soc_addr;
+  memset (&soc_addr, '\0', sizeof (struct sockaddr_ll));
+  prepare_sockaddr_ll (&soc_addr, ifindex, dst_haddr);
+
+  /* Init capture */
+  if (use_pcap != 0 && bpf < 0)
+    {
+      struct in_addr sin, this_host;
+      memset (&sin, '\0', sizeof (struct in_addr));
+      memset (&this_host, '\0', sizeof (struct in_addr));
+      if (IN6_IS_ADDR_V4MAPPED (ipaddr))
+        {
+          sin.s_addr = ipaddr->s6_addr32[3];
+          bpf = init_capture_device (sin, this_host, filter);
+        }
+      else
+        nasl_perror (
+          lexic, "%s: Error. Only IPv4 is supported for starting a capture.",
+          __func__);
+    }
+
+  // Prepare the message and send it
+  memset (&message, '\0', sizeof (struct msghdr));
+  prepare_message (&message, &soc_addr, (u_char *) frame, frame_sz);
+
+  int b = sendmsg (soc, &message, 0);
+  if (b == -1)
+    nasl_perror (lexic, "%s: Error sending message: %s", __func__,
+                 strerror (errno));
+
+  if (bpf >= 0)
+    answer =
+      (u_char *) capture_next_frame (bpf, to, &answer_sz, frame_and_payload);
+
+  if (answer)
+    {
+      retc = alloc_typed_cell (CONST_DATA);
+      retc->x.str_val = (char *) answer;
+      retc->size = answer_sz;
+    }
+
+  if (bpf >= 0)
+    bpf_close (bpf);
+  close (soc);
+
+  return retc;
+}
+
+/** @brief Dump a datalink layer frame
+ *
+ * @naslparam
+ *
+ * - @n frame The frame to be dumped.
+ *
+ * @param lexic Lexical context of NASL interpreter.
+ *
+ * @return Null
+ */
+tree_cell *
+nasl_dump_frame (lex_ctxt *lexic)
+{
+  u_char *frame = (u_char *) get_str_var_by_name (lexic, "frame");
+  int frame_sz = get_var_size_by_name (lexic, "frame");
+  int f = 0;
+
+  if (frame_sz == 0)
+    return NULL;
+
+  printf ("\nThe Frame:\n");
+  while (f < frame_sz)
+    {
+      printf ("%02x%02x ", ((u_char *) frame)[f], ((u_char *) frame)[f + 1]);
+      f += 2;
+      if (f % 16 == 0)
+        printf ("\n");
+    }
+  printf ("\n\n");
```

Comment:
Why using `printf`?

Reply:
Because for dumping in the console we use printf. It more for nasl developers. We use it also in other places like nasl_packet_forgery.c